libs/common/src/dtfs-central-api-response-types/utilisation-reports.ts

```diff
@@ -0,0 +1,9 @@
+import { ReportPeriod, IsoDateTimeStamp, AzureFileInfo, UtilisationReportEntity, Prettify } from '..';
+
+export type UtilisationReportResponseBody = Prettify<
```

Comment:
I'm team no link, I think, but could be swayed to a convincing argument. My thought process is mainly about avoiding accidental exposure of new data on endpoints:

I'd expect our api-tests to validate the received response against a hardcoded object so we could pick up if there were changes to the entity type that was affecting our response type. That being said, we should also not rely on this.

So the question is -- what's the benefit of linking dtos to our db entities (assuming ive scanned this right)?

Well, for one, we should not use `omit` for the reasons you've stated -- the entity type can grow, and we can't rely on our tests to be written consistently enough pick this up.

We can use `Pick` relatively fine. However, that's not entirely correct. if `Pick` picks a nested type, we are in the exact same position. Therefore `Pick` can only be used on basic types.

So why would we even include picking from an entity, and not just type it manually ourselves to avoid potential accidents here (its easier to see **never** linking to an entity rather than **sometimes** its okay to)?

However -- where do we stop here -- does everything, such as `AzureFileInfo`, have the scope to grow past what we want our response to contain?  Do we draw the line with not linking `entities` to `responses`, or do we go further and insist that each `response` must only use primitives and other `response` objects?

Reply:
Also -- A side point as I've not had eyes on this area, are we ensuring that at runtime we are stripping out fields that we don't want to reveal outside the service (because if we aren't -- depending on how we create the response object, we might be returning things we dont want to get out anyway)?

Reply:
Happy to change it back to how I previously wrote it if the consensus is we are happy with no linking :) I do personally prefer it as a general pattern.

@AlexBramhill I don't believe we are stripping anything out in dtfs-central-api but we are only actually using a few fields in portal-api and tfm api. I don't see there being any particular danger in things getting out, but I can double check this type captures all the returned fields.

Reply:
have a quick check that @ChristianMcc is happy with my mad ramblings :) 

Ignore the comment about what we are returning -- i've raised this offline to discuss if we want any endpoint specific middlewear for it

Reply:
I feel like i've flipped flopped between both approaches multiple times. I think I'll lean towards no overlapping similar again because we're the ones that are controlling both sides of the API so that's not too much of an issue in terms of missing parameters as our stuff will break if we don't provide the right information! 
It's not an area i've done a tonne of reading about so I'm basing that more on gut feel however than 'this is best practice'

Reply:
To further bloat this comment thread, I'm not sure we should have these types in `libs/common`. I think response types should be defined in each services `api.js` file as a `JSDoc` which can then be imported in typescript files as they are needed (or in the controller depending on which side of the request the service is on). If we define them here, things may get quite confusing when responses are passed through the various apis, but defining them at the service level next to the actual `api.js` function itself will make it obvious that this response type refers specifically to this api endpoint

Reply:
Hmm we have plenty of stuff in common that is meant to be consumed by only several services. The idea was since portal-api and trade-finance-manager-api both call the same endpoint by utilising the common library we only have one type we need to keep in line with what dtfs returns instead of two and them get out of sync, but happy to duplicate the type across the apis if I am slightly abusing the use of the common library here. Still trying to get my head around the desired patterns with regards to this common library 😄

Reply:
My main worry with it is that we often don't do anything extra to the response in `portal-api` or `tfm-api` (at least with the work we do), so we would have to define the type multiple times no matter what we do. It might just be a nicer pattern to try and keep type definitions for things like responses as close as possible to where the function is called I think

Reply:
Hadn't even seen these newer replies when I started the call earlier, I now see why we started talking about this part more and more :D
i'm a little against defining JSDoc types and then importing into typescript, I think it's just one extra quite fiddly thing to do and we have so much inside these api.js files already that's it's going to take us so long. I think I'd just prefer us to work with types in typescript where we can and leave simple JSDocs on new functions where approrpriate in javascript. 

In terms of 'common library or individual microservice' I'd mostly treat it as:
if sql entity, common library
if type is generic to DTFS rather than an individual microservice, common library
if type is only to be used in single microservice, put it in the micorservice or common library if there is a good place for it (I'm saying use your best judgement here/it doesn't matter, in typescript we'd import our types nicely enough that you can always find the type you're looking for where it's used so it doesn't matter too much)
Personally I think it's more microservice level models than response types that fit this kind of bill